### DISCUSSION TITLE
Request - Docker Compose Stack

### DISCUSSION THREAD

**Discussion by tomtom215** [Payroll-Engine/PayrollEngine#2] [Ideas]
Hello!

This project looks very impressive and fits the need for a lot of use cases.

Would it be possible for you to containerize this and publish an official docker image?

I think this will attract a much bigger audience to the project and allow people to quickly spin up something locally to test out.

**Comment by Giannoudis** [comment]
Thank you very much for your contribution.

I have also been thinking about offering it as a Docker image. I will tackle your idea shortly.

**Comment by Giannoudis** [comment]
I have made a sketch of the possible docker composition:
https://github.com/Payroll-Engine/PayrollEngine/blob/main/images/DockerMap.png
For the database, there is a T-SQL setup/update script that is installed using the Db Query Tool.

What is the most useful docker composition for production use?

**Comment by gsayer** [comment]
👋 **Hello!**

I propose the following modifications for a simplified `docker-compose` setup in the [`/PayrollEngine`](https://github.com/Payroll-Engine/PayrollEngine/pull/6) repository:

- [`PayrollEngine.Backend`](https://github.com/Payroll-Engine/PayrollEngine.Backend/pull/2)  
- [`PayrollEngine.WebApp`](https://github.com/Payroll-Engine/PayrollEngine.WebApp/pull/1)  
- [`PayrollEngine.SqlServer.DbQuery`](https://github.com/Payroll-Engine/PayrollEngine.SqlServer.DbQuery/pull/1)  
- [`PayrollEngine`](https://github.com/Payroll-Engine/PayrollEngine/pull/6)

---

**Notes:**

1. The 4 repositories must be placed at the same level in the folder hierarchy so that the images can be built correctly using `docker-compose`.
2. The `docker-compose` file launches a `db-init` container, which initializes the database and then stops itself.

---

**Next step:** We could automate the image builds using GitHub Actions.

---

@Giannoudis – I'm currently working on an embedded payroll platform project, so I don’t have full visibility on yours yet, but it looks incredible what you’ve built so far. Really impressive work!

**Comment by Giannoudis** [comment]
@gsayer  - Welcome to the Payroll Engine Community!  Thank you very much for your collaboration.  👍

I tested your extensions and found the following:

1. The `db-init` action does not work if I use a different password in the `.env` file. An error message stating that the password did not match the one provided appears.
```txt
2025-07-11 09:40:29.88 spid42s     Recovery is complete. This is an informational message only. No user action is required.
payrollengine-db-init  | Creating database...
2025-07-11 09:40:44.63 Logon       Error: 18456, Severity: 14, State: 8.
2025-07-11 09:40:44.63 Logon       Login failed for user 'sa'. Reason: Password did not match that for the login provided. [CLIENT: 172.18.0.3]
```

2. When I run 'dotnet publish' with the '--no-restore' option, I receive an error message saying that the NuGets cannot be resolved. This works without this option. Please note that the `C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages` package fallback folder exists locally.

```txt
 => ERROR [backend-api build 20/20] RUN dotnet publish "PayrollEngine.Backend.Server.csproj" -c Release -o /app/publish --no-restore                                                                                                                                                        1.1s
------
 > [backend-api build 20/20] RUN dotnet publish "PayrollEngine.Backend.Server.csproj" -c Release -o /app/publish --no-restore:
1.051 /usr/share/dotnet/sdk/9.0.302/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(266,5): error MSB4018: The "ResolvePackageAssets" task failed unexpectedly. [/src/Backend.Server/PayrollEngine.Backend.Server.csproj]
1.051 /usr/share/dotnet/sdk/9.0.302/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(266,5): error MSB4018: NuGet.Packaging.Core.PackagingException: Unable to find fallback package folder 'C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages'. [/src/Backend.Server/PayrollEngine.Backend.Server.csproj]
```

3.  I cannot log in to the database on the server `localhost:1433` using the `sa/.env-password` account in my SQL Server Management Console. Access to the backend and web app servers works fine.

4. The `PayrollEngine.SqlServer.DbQuery` tool is not needed anymore, the Container support is not required.

**Comment by Giannoudis** [comment]
@gsayer  - Thank your for the update. 

I have now the following database connection error in `db-init`:
```txt
Waiting for database to be ready...
/bin/sh: 5: /opt/mssql-tools18/bin/sqlcmd: not found
Testing database connection...
Database connection failed
```

The error occurs with every admin password.

**Reply by gsayer** [reply]
@Giannoudis

I think that the error `/bin/sh: 5: /opt/mssql-tools18/bin/sqlcmd: not found` is misleading. The real cause is probably the `!` character in your SQL Server password causing authentication failures that get masked as "command not found" errors.

## Solution

Fix the password in `.env`:
```bash
# Remove the ! character
DB_PASSWORD=XXXXX
```

Reset Docker environment:
```bash
docker-compose down
docker volume rm payrollengine_mssql-data
docker-compose up -d
```

[Commit c2f25b1](https://github.com/gsayer/PayrollEngine/commit/c2f25b183c8847c6a98ed8371962a7ad3ca61fdf)

**Comment by Giannoudis** [comment]
@gsayer

Thanks for the update. Even after fixing the password and resetting Docker, I'm still getting the `/bin/sh: 5: /opt/mssql-tools18/bin/sqlcmd: not found` error 30 seconds later.

**Reply by gsayer** [reply]
I can't reproduce the error when running db-init.

### My scenario
#### I have done:
```
docker-compose down
docker volume rm payrollengine_mssql-data
echo "DB_PASSWORD=PayrollStrongPass789" > .env
docker-compose up --build
```

#### My terminal output (container payrollengine-db-init):
```
payrollengine-db-init  | Testing database connection...
payrollengine-db-init  |                                                                                                                                                                                                                                                                                                             
payrollengine-db-init  | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
payrollengine-db-init  | Microsoft SQL Server 2022 (RTM-CU20) (KB5059390) - 16.0.4205.1 (X64) 
payrollengine-db-init  |        Jun 13 2025 13:38:45 
payrollengine-db-init  |        Copyright (C) 2022 Microsoft Corporation
payrollengine-db-init  |        Developer Edition (64-bit) on Linux (Ubuntu 22.04.5 LTS) <X64>                                                                                                      
payrollengine-db-init  | 
payrollengine-db-init  | (1 rows affected)
payrollengine-db-init  | Creating database...
2025-07-16 17:33:43.71 spid62      [5]. Feature Status: PVS: 0. CTR: 0. ConcurrentPFSUpdate: 1. ConcurrentGAMUpdate: 1. ConcurrentSGAMUpdate: 1, CleanupUnderUserTransaction: 0. TranLevelPVS: 0
2025-07-16 17:33:43.71 spid62      Starting up database 'PayrollEngine'.
2025-07-16 17:33:43.71 spid62      RemoveStaleDbEntries: Cleanup of stale DB entries called for database ID: [5]
2025-07-16 17:33:43.71 spid62      RemoveStaleDbEntries: Cleanup of stale DB entries skipped because master db is not memory optimized. DbId: 5.
2025-07-16 17:33:43.74 spid62      Parallel redo is started for database 'PayrollEngine' with worker pool size [5].
2025-07-16 17:33:43.74 spid62      Parallel redo is shutdown for database 'PayrollEngine' with worker pool size [5].
payrollengine-db-init  | Executing SQL script...
payrollengine-db-init  | The module 'DeleteAllCaseValues' depends on the missing object 'dbo.DeleteAllGlobalCaseValues'. The module will still be created; however, it cannot run successfully until the object exists.
payrollengine-db-init  | The module 'DeleteAllCaseValues' depends on the missing object 'dbo.DeleteAllNationalCaseValues'. The module will still be created; however, it cannot run successfully until the object exists.
payrollengine-db-init  | The module 'DeleteAllCaseValues' depends on the missing object 'dbo.DeleteAllCompanyCaseValues'. The module will still be created; however, it cannot run successfully until the object exists.
payrollengine-db-init  | The module 'DeleteAllCaseValues' depends on the missing object 'dbo.DeleteAllEmployeeCaseValues'. The module will still be created; however, it cannot run successfully until the object exists.
payrollengine-db-init  | 
payrollengine-db-init  | (1 rows affected)
payrollengine-db-init  | Payroll Engine database version successfully updated to release 0.9.2
payrollengine-db-init  | Database initialization completed.
payrollengine-db-init exited with code 0
```

### Check

```
docker exec payroll-db /opt/mssql-tools18/bin/sqlcmd -S localhost -U sa -P "PayrollStrongPass789" -C -Q "SELECT @@VERSION"
```

#### Output
 ```                                                                                                                                                                                                                                                                                                          
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Microsoft SQL Server 2022 (RTM-CU20) (KB5059390) - 16.0.4205.1 (X64) 
        Jun 13 2025 13:38:45 
        Copyright (C) 2022 Microsoft Corporation
        Developer Edition (64-bit) on Linux (Ubuntu 22.04.5 LTS) <X64>                                                                                                      

(1 rows affected)
```

**Reply by Giannoudis** [reply]
It appears to be a quoting issue. Setting the password without quotation marks now works on my system (Win10Pro).
```sh
docker-compose down
docker volume rm payrollengine_mssql-data
echo DB_PASSWORD=PayrollStrongPass789 > .env
docker-compose up --build
```

**Comment by Giannoudis** [comment]
@gsayer

Your merge requests have now been incorporated into the main branches. There is now a wiki page [Container Setup](https://github.com/Payroll-Engine/PayrollEngine/wiki/ContainerSetup) that explains how to use the Payroll Engine with Docker.

Thank you very much for your contribution — it is much appreciated!

@tomtom215 Thanks to @gsayer, the Payroll Engine now also supports Docker images.
